ODIN_II/SRC/netlist_cleanup.cpp

```
@@ -237,11 +242,81 @@ void calculate_addsub_statistics(node_list_t* addsub) {
     /* Calculate the geometric mean carry chain length */
     geomean_addsub_length = exp(sum_of_addsub_logs / total_addsub_chain_count);
 }
+void count_node_type(nnode_t* node) {
+    switch (node->type) {
+        // fallthrough
+        case LOGICAL_OR:
+        case LOGICAL_AND:
+        case LOGICAL_NOR:
+        case LOGICAL_NAND:
+        case LOGICAL_XOR:
+        case LOGICAL_XNOR:
+        case LOGICAL_NOT:
+        case LOGICAL_EQUAL:
+        case NOT_EQUAL:
+        case BITWISE_NOT:
+        case BUF_NODE:
+        case LT:
+        case GT:
+        case FF_NODE:
+        case MEMORY:
+        case MUX_2:
+        case SMUX_2:
+        case MINUS:
+        case ADD:
+        case MULTIPLY:
+        case HARD_IP:
+        case INPUT_NODE:
+        case OUTPUT_NODE:
+        case PAD_NODE:
+        case GND_NODE:
+        case VCC_NODE:
+        case CLOCK_NODE:
+        case ADDER_FUNC:
+        case CARRY_FUNC:
+            num_removed_nodes[node->type]++;
+            break;
+        case BITWISE_AND:
+        case BITWISE_NAND:
+        case BITWISE_NOR:
+        case BITWISE_XNOR:
+        case BITWISE_XOR:
+        case BITWISE_OR:
+        case MULTI_PORT_MUX:
+        case SL:
```

Why shift nodes should throw an error here? As I mentioned last time please have a look at the big switch-case in ```simulate_blif.cpp:771 => compute_and_store_value```. You would need to check such cases. You need to throw an error ONLY for operation types that are already resolved into other logic, such as bitwise operations, MULTI_PORT_MUX, CASE_EQUAL, GTE, ... .

---

If I'm correct, I believe that you mentioned to use the case statement in  `BLIFWriter.cpp:458`. In `BLIFWriter` case statement, shift nodes throw an error.

---

Either of them should work fine; the point is to be aware that some nodes may not be resolved into the fine-grained level(ready for output) yet at this stage, so you would need to take care of those cases. That's actually why we see a bunch of errors for the Odin-II regression test (e.g. you throw an error for arithmetic shift operation here why it shouldn't be so). So I would recommend verifying the Odin-II functionality for at least the micro or syntax regression test before pushing a commit so you would be able to see errors clearly.
```
./verify_odin.py -t regression_test/benchmarks/task/micro
```

---

Thanks, will do.